src/Sylius/Bundle/ResourceBundle/Storage/CookieStorage.php

```diff
@@ -0,0 +1,124 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\ResourceBundle\Storage;
+
+use Sylius\Component\Resource\Storage\StorageInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpFoundation\Cookie;
+use Symfony\Component\HttpFoundation\ParameterBag;
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+
+/**
+ * @author Kamil Kokot <kamil.kokot@lakion.com>
+ */
+final class CookieStorage implements StorageInterface, EventSubscriberInterface
```

Comment:
The one and only responsibility of cookie storage is to allow for accessing and storing values in browser's cookies - which we can get after the request is sent and set just before response is sent back.

Dividing this class into two separate ones will result in implementation leaks (we only set these cookies that aren't already set and store the existing ones separately, it's not "**just** updating"), difficulties in connecting these two together (as the cookie listener is **only** applicable to cookie storage). All of them can be used separately but it would not work correctly until they are used together.